File: problems/0010_regular-expression-matching/backtrace/solution.rb
#!/usr/bin/ruby

# @param {String} s
# @param {String} p
# @return {Boolean}
def is_match(s, p)
    if p.length == 0
        return (s.length == 0)
    end
    p0 = p[0]
    match_p0 = (s.length > 0 and
            (p0 == s[0] or p0 == '.'))
    if p.length >= 2 and p[1] == '*'
        return (is_match(s, p[2..-1]) or
                (match_p0 and is_match(s[1..-1], p)))
    else
        return (match_p0 and is_match(s[1..-1], p[1..-1]))
    end
end

File: problems/0010_regular-expression-matching/backtrace/Solution.py
#!/usr/bin/python3

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        sLen, pLen = len(s), len(p)
        if pLen == 0: return (sLen == 0)
        p0 = p[0]
        matchP0 = (sLen > 0 and \
                (p0 == s[0] or p0 == '.'))
        if pLen >= 2 and p[1] == '*':
            return (self.isMatch(s, p[2:]) or \
                    (matchP0 and self.isMatch(s[1:], p)))
        else:
            return (matchP0 and self.isMatch(s[1:], p[1:]))
